django_walletpass/models.py

```diff
@@ -308,12 +314,83 @@ class Registration(models.Model):
+    created_at = models.DateTimeField(default=timezone.now)
+    status = models.CharField(max_length=100, null=True, blank=True)
+    task_type = models.CharField(max_length=255, null=True, blank=True)
+    pass_type_identifier = models.CharField(max_length=255, null=True, blank=True)
+    serial_number = models.CharField(max_length=255, null=True, blank=True)
+    pazz = models.ForeignKey(Pass, null=True, blank=True, on_delete=models.CASCADE, related_name='logs')
+    web_service_url = models.URLField(null=True, blank=True)
+    device_id = models.CharField(max_length=255, null=True, blank=True)
+    msg = models.TextField(null=True, blank=True)
     message = models.TextField()
 
     def __unicode__(self):
         return self.message
+
+    def __str__(self):
+        return self.created_at.strftime('%d/%m/%y %H:%M:%S')
+
+    @classmethod
+    def parse_log(cls, log, message):
+        pattern_register = r"\[(.*?)\]\s(.*?)\s\(for device (.*?), pass type (.*?), serial number (.*?); with web service url (.*?)\)\s(.*?): (.*$)"
+        pattern_get = r"\[(.*?)\]\s(.*?)\s\(pass type (.*?), serial number (.*?), if-modified-since \(.*?\); with web service url (.*?)\) (.*?): (.*$)"
+        pattern_web_service_error = r"\[(.*?)\]\s(.*?)\sfor (.*?)\s\((.*?)\):\s(.*$)"
+        pattern_get_warning = r"\[(.*?)\]\s(.*?)\s\(pass type (.*?), serial number (.*?), if-modified-since \(.*?\); with web service url (.*?)\) (.*?): (.*\.)\s(.*$)"
+
+        match_register = re.match(pattern_register, message)
+        match_get = re.match(pattern_get, message)
+        match_web_service_error = re.match(pattern_web_service_error, message)
+        match_get_warning = re.match(pattern_get_warning, message)
+
+        if match_register:
+            timestamp_str, task_type, device_id, pass_type_identifier, serial_number, web_service_url, status, msg = match_register.groups()
+        elif match_get:
+            timestamp_str, task_type, pass_type_identifier, serial_number, web_service_url, status, msg = match_get.groups()
+            device_id = None  # 'Get pass task' entries don't include device_id
+        elif match_web_service_error:
+            timestamp_str, task_type, pass_type_identifier, web_service_url, msg = match_web_service_error.groups()
+            serial_number = None
+            device_id = None
+            status = "error"
+        elif match_get_warning:
+            timestamp_str, task_type, pass_type_identifier, serial_number, web_service_url, status, msg = match_get_warning.groups()
+            device_id = None
+            status = "warning"
+        else:
+            log.status = 'unknown'
+            log.message = message
+            log.save()
+            return  # Log entry didn't match any known pattern
+
+        if 'error' in status:
+            status = 'error'
+        elif 'warning' in status:
+            status = 'warning'
+
+        log.created_at = datetime.datetime.strptime(timestamp_str, "%Y-%m-%d %H:%M:%S %z")
```

Comment:
This will crash if `timestamp_str` doesn't have the expected format (which will happen if the code falls through the final `else`). Maybe default to `now()`?

Reply:
good morning, the final `else` will trigger return. in that case `timestamp_str` wouldn't even be defined. in case `timestamp_str` changes and does not match the expected pattern anymore, we should get noticed (exception?) and not fill it with a default value which would result in the data speaking non-exact facts.

Reply:
ohh, true! hmmm... 🤔... maybe add a new TextField and store the raw received message? This way we won't loose any messages (that might be useful, but we didn't have the right pattern match for them).

Reply:
like that: https://github.com/Develatio/django-walletpass/pull/29/files#diff-e6bb9459442a32d45360786af0d351dded535339061352e4ee77886cd155e48cR370 ?

Reply:
😅 time for another coffee I guess